### PR TITLE
fix prototype transforms tests with set agg_method

### PR DESCRIPTION
**Pull request by pmeier**
`assert_{close, equal}` was buggy for prototype tests if one set the `agg_method` as we do for a few tests through

https://github.com/pytorch/vision/blob/f32600b66d9c03682ccd975e442e67c609fd7f0a/test/prototype_transforms_kernel_infos.py#L64-L67

I only wanted to fix that, but realized fixing this entails a lot more. A lot of the reference test that used these tolerances failed. Thus, this PR only lays the foundation to fix them in the near future.

cc @vfdev-5 @datumbox @bjuncek

### CODE REVIEW

File: test/prototype_common_utils.py [hunk 1]
```diff
@@ -111,16 +72,24 @@ def compare(self) -> None:
...
 
         if self.agg_method is None:
             super()._compare_values(actual, expected)
         else:
-            err = self.agg_method(abs_diff.to(torch.float64))
-            if err > self.atol:
-                self._make_error_meta(AssertionError, "aggregated mismatch")
+            agg_abs_diff = float(self.agg_method(abs_diff.to(torch.float64)))
+            if agg_abs_diff > self.atol:
+                raise self._make_error_meta(
```
Comment (pmeier) [inline test/prototype_common_utils.py:88]: This was the source of the actual bug. Without the raise we just created an exception, but never did anything with it. Thus, all tests that set `agg_method` in their `closeness_kwargs` passed without a value check.

File: test/prototype_common_utils.py [hunk 2]
```diff
@@ -54,7 +46,7 @@
 ]
 
 
-class PILImagePair(TensorLikePair):
+class ImagePair(TensorLikePair):
```
Comment (pmeier) [inline test/prototype_common_utils.py:49]: I've refactored this to not handle mixed PIL / tensor image pairs. It was a problem if the tolerance is set for floating point images, i.e. in the range `[0.0, 1.0]`, but the comparison converted to uint8, which needs higher tolerances.

File: test/prototype_transforms_kernel_infos.py [hunk 3]
```diff
@@ -62,8 +63,8 @@ def __init__(
 
 
 DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS = {
-    (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(atol=1e-5, rtol=0, agg_method="mean"),
-    (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(atol=1e-5, rtol=0, agg_method="mean"),
+    (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(atol=0.9, rtol=0, agg_method="mean"),
+    (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(atol=255 * 0.9, rtol=0, agg_method="mean"),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:67]: These are crazy tolerances and render the test effectively useless. This PR just lays the foundation to fix these in the near future. I'll open an issue ASAP with a game plan to review all the operators again that need these tolerances for some reason.

File: test/prototype_transforms_kernel_infos.py [hunk 4]
```diff
@@ -74,14 +75,26 @@ def __init__(
...
             )
 
-        # We don't need to convert back to tensor here, since `assert_close` does that automatically.
-        return pil_kernel(F.to_image_pil(image_tensor), *other_args, **kwargs)
+        input_pil = F.to_image_pil(input_tensor)
+        output_pil = pil_kernel(input_pil, *other_args, **kwargs)
+        if not isinstance(output_pil, PIL.Image.Image):
+            return output_pil
+
+        output_tensor = F.convert_dtype_image_tensor(F.to_image_tensor(output_pil), dtype=input_tensor.dtype)
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:89]: Since there is no more implicit conversion in the comparison logic, we need to convert here. This gives us the ability to set the correct dtype by using the one input is also using.

File: test/prototype_transforms_kernel_infos.py [hunk 5]
```diff
@@ -74,14 +75,26 @@ def __init__(
...
+
+        # 2D mask shenanigans
+        if output_tensor.ndim == 2 and input_tensor.ndim == 3:
+            output_tensor = output_tensor.unsqueeze(0)
+        elif output_tensor.ndim == 3 and input_tensor.ndim == 2:
+            output_tensor = output_tensor.squeeze(0)
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:95]: For all other shape mismatches we can let the comparison logic handle the error.

File: test/prototype_transforms_kernel_infos.py [hunk 6]
```diff
@@ -400,6 +413,23 @@ def _full_affine_params(**partial_params):
 ]
 
 
+def _get_fills(*, num_channels, dtype, vector=True):
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:416]: `fill` was creating issues because we set something like `128.0` for floating point images. This is not an issue in general, but affects stuff like aggregated diff. This just generalizes the fill generation for all tests below.

File: test/prototype_transforms_kernel_infos.py [hunk 7]
```diff
@@ -631,7 +658,9 @@ def reference_inputs_convert_format_bounding_box():
 
 
 def sample_inputs_convert_color_space_image_tensor():
-    color_spaces = list(set(features.ColorSpace) - {features.ColorSpace.OTHER})
+    color_spaces = sorted(
+        set(features.ColorSpace) - {features.ColorSpace.OTHER}, key=lambda color_space: color_space.value
+    )
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:663]: Drive-by that I found while debugging. Without this, individual failing tests are not reproducible because the order is not stable.

File: test/prototype_transforms_kernel_infos.py [hunk 8]
```diff
@@ -678,7 +707,10 @@ def sample_inputs_convert_color_space_video():
             sample_inputs_fn=sample_inputs_convert_color_space_image_tensor,
             reference_fn=reference_convert_color_space_image_tensor,
             reference_inputs_fn=reference_inputs_convert_color_space_image_tensor,
-            closeness_kwargs=DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+            closeness_kwargs={
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(atol=2 / 255, rtol=0),
+                (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(atol=1, rtol=0),
+            },
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:713]: I already fixed the tolerances here, because I was debugging the issue above. Fortunately, tolerances are fine.

File: test/prototype_transforms_kernel_infos.py [hunk 9]
```diff
@@ -2070,6 +2099,17 @@ def sample_inputs_ten_crop_video():
         yield ArgsKwargs(video_loader, size=size)
 
 
+def multi_crop_pil_reference_wrapper(pil_kernel):
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:2102]: Small helper since the regular `pil_reference_wrapper` cannot handle conversion of these tuple or list outputs.

File: test/test_prototype_transforms_consistency.py [hunk 10]
```diff
@@ -1005,7 +1005,7 @@ def make_datapoints(self, supports_pil=True, image_dtype=torch.uint8):
 
             dp = (conv_fn(feature_image), feature_mask)
             dp_ref = (
-                to_image_pil(feature_image) if supports_pil else torch.Tensor(feature_image),
+                to_image_pil(feature_image) if supports_pil else feature_image.as_subclass(torch.Tensor),
```
Comment (pmeier) [inline test/test_prototype_transforms_consistency.py:1008]: Driveby.

File: test/test_prototype_transforms_consistency.py [hunk 11]
```diff
@@ -1019,12 +1019,16 @@ def check(self, t, t_ref, data_kwargs=None):
...
-            output = t(dp)
+            actual = actual_image, actual_mask = t(dp)
 
             self.set_seed()
-            expected_output = t_ref(*dp_ref)
+            expected_image, expected_mask = t_ref(*dp_ref)
+            if isinstance(actual_image, torch.Tensor) and not isinstance(expected_image, torch.Tensor):
+                expected_image = legacy_F.pil_to_tensor(expected_image)
+            expected_mask = legacy_F.pil_to_tensor(expected_mask).squeeze(0)
+            expected = (expected_image, expected_mask)
```
Comment (pmeier) [inline test/test_prototype_transforms_consistency.py:1029]: This was relying on the mixed PIL / tensor image comparison that was removed above. Thus, we do it manually here.

File: test/test_prototype_transforms_functional.py [hunk 12]
```diff
@@ -237,7 +237,6 @@ def test_against_reference(self, test_id, info, args_kwargs):
         assert_close(
             actual,
             expected,
-            check_dtype=False,
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:240]: We can be strict here now, since we perform the conversion correctly.

File: test/test_prototype_transforms_consistency.py (hunk 10 above)
Comment (vfdev-5) [reply test/test_prototype_transforms_consistency.py:1008]: micro optimization :)

File: test/prototype_transforms_kernel_infos.py (hunk 3 above)
Comment (datumbox) [reply test/prototype_transforms_kernel_infos.py:67]: These are crazy tolerances indeed. From your earlier comment, I understand that the tests were effectively not throwing exceptions, so here you adjust the values to make them pass and then revisit all kernels that fail. Is my understanding correct?

**Comment by datumbox** [comment]
@pmeier I think it's worth fixing the issues in this PR and merging later. What's the purpose of making a bunch of test corrections and then disabling the tests? No strong opinion though.

File: test/prototype_transforms_kernel_infos.py (hunk 3 above)
Comment (pmeier) [reply test/prototype_transforms_kernel_infos.py:67]: Yes.

**Review by datumbox — APPROVED** [review]
Offering a stamp to unblock the work. Up to you @pmeier if you want to make the changes here to make all tests pass or do it gradually on follow ups. My only concern is turning off all references might be dangerous. Perhaps we could skip only those that fail?

File: test/prototype_transforms_kernel_infos.py [hunk 13]
```diff
@@ -325,7 +335,14 @@ def reference_inputs_resize_bounding_box():
...
             closeness_kwargs={
-                **DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+                # TODO: These tolerances are noticeably larger than "regular" ones. Investigate if this is a bug in the
+                #  kernel.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(
+                    atol=110 / 255, rtol=0, agg_method="mean"
+                ),
+                (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(
+                    atol=110, rtol=0, agg_method="mean"
+                ),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:345]: `resize_image_tensor`

File: test/prototype_transforms_kernel_infos.py [hunk 14]
```diff
@@ -564,7 +597,14 @@ def sample_inputs_affine_video():
...
             reference_inputs_fn=reference_inputs_affine_image_tensor,
-            closeness_kwargs=DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+            closeness_kwargs={
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(
+                    atol=10 / 255, rtol=0, agg_method="mean"
+                ),
+                (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(
+                    atol=10, rtol=0, agg_method="mean"
+                ),
+            },
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:607]: A average diff of `10` is still pretty high, but far from the other problematic ones. We can still investigate though.

File: test/prototype_transforms_kernel_infos.py [hunk 15]
```diff
@@ -830,7 +878,16 @@ def sample_inputs_rotate_video():
...
-            closeness_kwargs=DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+            closeness_kwargs={
+                # TODO: These tolerances are noticeably larger than "regular" ones. Investigate if this is a bug in the
+                #  kernel.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(
+                    atol=120 / 255, rtol=0, agg_method="mean"
+                ),
+                (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(
+                    atol=90, rtol=0, agg_method="mean"
+                ),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:889]: `rotate_image_tensor`

File: test/prototype_transforms_kernel_infos.py [hunk 16]
```diff
@@ -1021,7 +1081,14 @@ def sample_inputs_resized_crop_video():
...
             closeness_kwargs={
-                **DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+                # TODO: These tolerances are noticeably larger than "regular" ones. Investigate if this is a bug in the
+                #  kernel.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(
+                    atol=60 / 255, rtol=0, agg_method="mean"
+                ),
+                (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(
+                    atol=60, rtol=0, agg_method="mean"
+                ),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:1091]: `resized_crop_image_tensor`. This can be resolved together with `resize_image_tensor`, since `crop_image_tensor` is fine.

File: test/prototype_transforms_kernel_infos.py [hunk 17]
```diff
@@ -1158,7 +1227,12 @@ def reference_inputs_pad_bounding_box():
             sample_inputs_fn=sample_inputs_pad_image_tensor,
             reference_fn=pil_reference_wrapper(F.pad_image_pil),
             reference_inputs_fn=reference_inputs_pad_image_tensor,
-            closeness_kwargs=DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+            closeness_kwargs={
+                # FIXME: This tolerance effectively renders the test useless since it cannot fail. We need to
+                #  investigate for what configuration this is happening. Since uint8 works perfectly there seems to be
+                #  a bug in the kernel.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(atol=1, rtol=0),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:1234]: `pad_image_tensor`. This is the worst one. `atol=1` for floating point inputs means any value inside the valid range of `[0.0, 1.0]` is fine.

File: test/prototype_transforms_kernel_infos.py [hunk 18]
```diff
@@ -1240,7 +1313,14 @@ def sample_inputs_perspective_video():
...
             closeness_kwargs={
-                **DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+                # TODO: This tolerance is noticeably larger than "regular" ones. Investigate if this is a bug in the
+                #  kernel.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(
+                    atol=160 / 255, rtol=0, agg_method="mean"
+                ),
+                (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(
+                    atol=160, rtol=0, agg_method="mean"
+                ),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:1323]: `perspective_image_tensor`

File: test/prototype_transforms_kernel_infos.py [hunk 19]
```diff
@@ -1324,7 +1406,13 @@ def sample_inputs_elastic_video():
             sample_inputs_fn=sample_inputs_elastic_image_tensor,
             reference_fn=pil_reference_wrapper(F.elastic_image_pil),
             reference_inputs_fn=reference_inputs_elastic_image_tensor,
-            closeness_kwargs=DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+            closeness_kwargs={
+                # TODO: This tolerance is noticeably larger than "regular" ones. Investigate if this is a bug in the
+                #  kernel for floating point inputs, given that uint8 works perfectly.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(
+                    atol=160 / 255, rtol=0, agg_method="mean"
+                ),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:1414]: `elastic_image_tensor`

File: test/prototype_transforms_kernel_infos.py [hunk 20]
```diff
@@ -1551,7 +1641,11 @@ def sample_inputs_equalize_video():
             sample_inputs_fn=sample_inputs_equalize_image_tensor,
             reference_fn=pil_reference_wrapper(F.equalize_image_pil),
             reference_inputs_fn=reference_inputs_equalize_image_tensor,
-            closeness_kwargs=DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+            closeness_kwargs={
+                # TODO: This tolerance is noticeably larger than "regular" ones. Investigate if this is a bug in the
+                #  kernel for floating point inputs, given that uint8 works perfectly.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(atol=12 / 255, rtol=0),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:1647]: `equalize_image_tensor`

File: test/prototype_transforms_kernel_infos.py [hunk 21]
```diff
@@ -1908,7 +2026,12 @@ def sample_inputs_adjust_gamma_video():
             sample_inputs_fn=sample_inputs_adjust_gamma_image_tensor,
             reference_fn=pil_reference_wrapper(F.adjust_gamma_image_pil),
             reference_inputs_fn=reference_inputs_adjust_gamma_image_tensor,
-            closeness_kwargs=DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+            closeness_kwargs={
+                # TODO: This tolerance is noticeably larger than "regular" ones. Investigate if this is a bug in the
+                #  kernel for floating point inputs, given that uint8 works almost perfectly.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(atol=40 / 255, rtol=0),
+                (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(atol=1, rtol=0),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:2033]: `adjust_gamma_image_tensor`. Since uint8 is fine, maybe this just happens due to our conversion to uint8 and back for PIL. Meaning, this could be a limitation of the test suite. Maybe it is sufficient to only check uint8 here against PIL and separately check that our kernel is consistent for float32 and uint8.

File: test/prototype_transforms_kernel_infos.py [hunk 22]
```diff
@@ -1949,7 +2072,12 @@ def sample_inputs_adjust_hue_video():
             sample_inputs_fn=sample_inputs_adjust_hue_image_tensor,
             reference_fn=pil_reference_wrapper(F.adjust_hue_image_pil),
             reference_inputs_fn=reference_inputs_adjust_hue_image_tensor,
-            closeness_kwargs=DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS,
+            closeness_kwargs={
+                # TODO: These tolerances are noticeably larger than "regular" ones. Investigate if this is a bug in the
+                #  kernel.
+                (("TestKernels", "test_against_reference"), torch.float32, "cpu"): dict(atol=20 / 255, rtol=0),
+                (("TestKernels", "test_against_reference"), torch.uint8, "cpu"): dict(atol=20, rtol=0),
+            },
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:2080]: `adjust_hue_tensor`

File: test/test_prototype_transforms_functional.py [hunk 23]
```diff
@@ -436,6 +439,29 @@ def test_dtype_and_device_convert_image_dtype(info, args_kwargs, device):
...
 
 
+@pytest.mark.parametrize(
+    ("info", "args_kwargs"),
+    make_info_args_kwargs_params(
+        KERNEL_INFOS_MAP[F.posterize_image_tensor],
+        args_kwargs_fn=lambda info: info.reference_inputs_fn(),
+    ),
+)
+def test_posterize_image_tensor_float_vs_int(info, args_kwargs):
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:449]: This test makes sure that `posterize_image_tensor` is consistent for uint8 vs float32. We need this since the conversion to PIL and back introduces enough noise that float32 cannot be compared without crazy tolerances.

Maybe we should do this for all image kernels: test only uint8 against PIL and have a separate test that checks dtype consistency for these kernels. Maybe we can avoid some of the tolerance madness above.

**Review by pmeier — COMMENTED** [review]
With the latest commit I removed the `DEFAULT_PIL_REFERENCE_CLOSENESS_KWARGS` in favor of setting the tolerances for each kernel separately. Some of them are pretty wild and I'm going to highlight them below. Given that everything works in our E2E tests, IMO this is either caused by parameter combinations that we don't use or a problematic test setup. 

We need to investigate them in any case. Let's use #6937 to coordinate. I'll update it soon to check everything that was already done in this PR.

**Review by datumbox — APPROVED** [review]
Still LGTM.

A few comments:
- We do know that accuracy is not affected but it might be the case that a kernel is broken/buggy on setups we haven't tested (floats for instance) or that the specific augmentation doesn't have much effect on accuracy. So as you highlighted we must check things very thoroughly to understand if this is a test problem or an actual bug.
- I wonder if it's worth it to focus on single pixel differences rather than averages. As averages can accumulate, it might worth examining if max single pixel tolerances can be stricter.
- Some of the tolerances look scary. We should bisect and revert any speed optimizations that cause them. It might be also worth checking our tensor kernels against PIL. Stable tests do that now and this is how we ensure that stable and PIL stay aligned. V2 doesn't have such tests (of course it compares itself with V1) but due to the increased tolerances we might be missing something important.

Let's merge on green CI and take this as top priority task to resolve.

**Comment by pmeier** [comment]
In the end I've refactored this by quite a bit so we get a clearer picture. Here are the main changes this PR brings:

- Remove the ability to check PIL vs. tensor images in `assert_close`. The automatic type and dtype conversion made it really hard to correctly specify tolerances.
- Remove the ability to perform reference tests against PIL with floating point tensors. Again, conversion automagic made it really hard to locate issues.
- Add a new test that checks image kernels for consistency between uint8 and float32. With this and the reference test against PIL for uint8, we have the same information as before, but are able to pinpoint problems easier.

With these changes, there are only a couple of operators with wild tolerances left to check:

- `adjust_hue_image_tensor`
- `elastic_image_tensor`
- `elastic_mask`
- `perspective_image_tensor`
- `resize_image_tensor`
- `resized_crop_image_tensor`
- `rotate_image_tensor`

They are tracked in #6937 and that will be used after this PR. 

Since these tests are pretty flaky right now, I cranked up the tolerances to avoid problems with the CI. Let's see if I need more.
